Subject: [PATCH] Allow bypassing check for host header.
NiFi has the configuration option 'nifi.web.proxy.host' which controls allowed
values for the host header field in any incoming request for the web ui.

This frequently causes issues when trying to expose the NiFi UI via for example
an ingress, loadbalancer or any similar type of mechanism.

NiFi does not allow to disable this behavior, so at the moment the nifi operator
simply hardcodes all even remotely possible values into this field.
But in order to allow putting for example in ingress in front of NiFi this means
using config overrides to change the value of this option, copy all the values
the operator put in there and add the extra value you need.

This is less than ideal, the proper solution would probably be
https://github.com/stackabletech/nifi-operator/issues/604

But until that is merged this is a simple workaround that allows overriding the list of allowed
hostnames by just setting it to "*" and this will effectively bypass the hostname check entirely if set.

This allows us to keep the default behavior in place for those users where it works and not remove
security features, but also enables users to disable this check if they know what they are doing.
---
Index: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/HostHeaderHandler.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/HostHeaderHandler.java b/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/HostHeaderHandler.java
--- a/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/HostHeaderHandler.java	(revision 6ecc398d3f92425447e43242af4992757e25b3c5)
+++ b/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/HostHeaderHandler.java	(date 1716453739677)
@@ -47,6 +47,7 @@
     private final String serverName;
     private final int serverPort;
     private final List<String> validHosts;
+    private boolean allowAllHosts = false;

     /**
      * Instantiates a handler with a given server name and port 0.
@@ -107,6 +108,10 @@
         // The value(s) from nifi.web.proxy.host
         hosts.addAll(parseCustomHostnames(niFiProperties));

+        // Check if the setting for allowed hosts has only the wildcard entry and
+        // if so store this in allowAllHost for later use
+        List<String> configuredHostNames = niFiProperties.getAllowedHostsAsList();
+        this.allowAllHosts = configuredHostNames.size() == 1 && configuredHostNames.contains("*");
         // empty is ok here
         hosts.add("");

@@ -205,7 +210,7 @@
     }

     boolean hostHeaderIsValid(String hostHeader) {
-        return validHosts.contains(hostHeader.toLowerCase().trim());
+        return this.allowAllHosts || validHosts.contains(hostHeader.toLowerCase().trim());
     }

     @Override
